File: algorithms/dp/best_time_to_buy_and_sell_stock_IV.rs
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock-iv/

// time complexity: O(kn)

use std::cmp::max;

impl Solution {
    pub fn max_profit(k: i32, prices: Vec<i32>) -> i32 {
        let k = k as usize;
        let (mut bought, mut sold) = (vec![i32::MIN; k + 1], vec![0; k + 1]);
        for p in &prices {
            for i in 1..k + 1 {
                sold[i] = max(sold[i], bought[i] + p);
                bought[i] = max(bought[i], sold[i - 1] - p);
            }
        }
        return sold[k];
    }
}

File: leetcode/best_time_to_buy_and_sell_stock_IV.py
# https://leetcode.com/problems/best-time-to-buy-and-sell-stock-iv/

# time complexity: O(kn)

class Solution:
    def maxProfit(self, k: int, prices: list[int]) -> int:
        bought, sold = [float('-inf')] * (k + 1), [0] * (k + 1)
        for p in prices:
            for i in range(1, k + 1):
                sold[i] = max(sold[i], bought[i] + p)
                bought[i] = max(bought[i], sold[i - 1] - p)
        return sold[-1]